### PR TITLE
Throttle profile fetch and retry on failure.

### PR DESCRIPTION
**Pull request by charlesmchen**
PTAL @michaelkirk

### CODE REVIEW

File: Signal/src/ProfileFetcherJob.swift [hunk 1]
```diff
@@ -26,30 +29,60 @@ class ProfileFetcherJob: NSObject {
...
+
+    public func getProfile(recipientId: String, remainingRetries: Int = 3) {
 
-                    Logger.error("\(self.TAG) failed to fetch profile for recipient: \(recipientId) with error: \(error)")
-            })
+        if let lastDate = ProfileFetcherJob.fetchDateMap[recipientId] {
+            let lastTimeInterval = fabs(lastDate.timeIntervalSinceNow)
+            // Don't check a profile more often than every N minutes.
+            let kGetProfileMaxFrequencySeconds = 60.0 * 5.0
+            if lastTimeInterval < kGetProfileMaxFrequencySeconds {
```
Comment (michaelkirk) [inline Signal/src/ProfileFetcherJob.swift]: style nit: use guard to enforce early-return pattern.

File: Signal/src/ProfileFetcherJob.swift [hunk 2]
```diff
@@ -26,30 +29,60 @@ class ProfileFetcherJob: NSObject {
...
+                return
+            }
         }
+        ProfileFetcherJob.fetchDateMap[recipientId] = Date()
+
+        Logger.error("\(self.TAG) getProfile: \(recipientId)")
+
+        let request = OWSGetProfileRequest(recipientId: recipientId)
+
+        // We don't need to retainUntilComplete() since the success and failure
```
Comment (michaelkirk) [inline Signal/src/ProfileFetcherJob.swift]: not sure I understand this comment - `retainUntilComplete` is a PromiseKit extension - but we're not using promises here.

File: Signal/src/ProfileFetcherJob.swift [hunk 3]
```diff
@@ -14,6 +14,9 @@ class ProfileFetcherJob: NSObject {
 
     let thread: TSThread
 
+    // This property is only accessed on the default global queue.  
+    static var fetchDateMap = [String: Date]()
```
Comment (charlesmchen) [inline Signal/src/ProfileFetcherJob.swift]: You're right. I've moved to main.

File: Signal/src/ProfileFetcherJob.swift (hunk 1 above)
Comment (charlesmchen) [inline Signal/src/ProfileFetcherJob.swift]: Done.

File: Signal/src/ProfileFetcherJob.swift (hunk 2 above)
Comment (charlesmchen) [inline Signal/src/ProfileFetcherJob.swift]: My thought process was that we tend to use promises in jobs and so will often want to retainUntilComplete() in jobs to the point where we should either a) use retainUntilComplete() or b) add a comment noting that it's not necessary.

But I guess it's excessive.  I'll remove it.

**Pull request merged**